container/archive.bzl

```diff
@@ -0,0 +1,110 @@
+# Copyright 2017 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Rule for importing an image from 'docker save' tarballs.
+
+This extracts the tarball, examines the layers and creates a
+container_import target for use with container_image.
+"""
+
+def _manifest_layers(manifest):
+  # Extracts layers, preserving order, from the manifest.
+
+  layers_begin_marker = "\"Layers\":["
+  start = manifest.index(layers_begin_marker)
+  end = manifest.index("]", start)
+
+  start += len(layers_begin_marker)
+  layers = manifest[start:end].replace("\"", "").split(",")
+
+  if not layers:
+    fail("Failed to extract layers from manifest.json")
+
+  return layers
+
+def _gzip_layers(ctx):
+  # Returns an array of gzipped layers, preserving order.
+
+  manifest = ctx.execute(["cat", "manifest.json"])
+  if manifest.return_code:
+    fail("Could not read manifest: %s" % manifest.stderr)
+
+  zipped_layers = []
+  layers = _manifest_layers(manifest.stdout)
+  for layer in layers:
+    zipped_layer = layer.split("/")[0] + ".tar.gz"
+    result = ctx.execute(["gzip", "-n", layer])
+    if result.return_code:
+      fail("Failed to gzip image layer %s: %s" % (layer, result.stderr))
+    result = ctx.execute(["mv", layer + ".gz", zipped_layer])
+    if result.return_code:
+      fail("Failed to gzip image layer %s: %s" % (layer, result.stderr))
+    zipped_layers += ["\"" + zipped_layer + "\","]
+
+  return zipped_layers
+
+def _config_json(ctx):
+  # Identifies the image config json file in the root directory.
+
+  config_candidates = ctx.execute(["ls", "-1"])
+  if config_candidates.return_code:
+    fail("Failed to list contents: %s" % config_candidates.stderr)
+
+  config = None
+  for candidate in config_candidates.stdout.splitlines():
+    if candidate.endswith(".json") and candidate != "manifest.json":
+      config = candidate
+      break
+
+  if not config:
+    fail("Failed to find config file")
+
+  return config
+
+def _container_archive_impl(ctx):
+  ctx.download_and_extract(url=ctx.attr.urls,
+                           sha256=ctx.attr.sha256,
+                           type=ctx.attr.type,
+                           stripPrefix=ctx.attr.strip_prefix)
+
+  zipped_layers = _gzip_layers(ctx)
+
+  config = _config_json(ctx)
```

Comment:
Turns out [I wrote a tool for this](https://github.com/google/containerregistry/blob/master/tools/fast_importer_.py#L46), but didn't add a `cloudbuild.yaml` to release the `PAR` file.

I'll do that, setup the build, and cut a release so we can pick that up.

Reply:
Looking through it now. Will update this PR when I am done.

Reply:
I tried it out. This PR is now using your tool. Thank you for pointing me to it and sending in the PR to include it.

However, I did notice that the tool is much slower than simply using gzip. For example, without using the importer tool, my image is imported in ~30s, but with the importer it takes ~5m30s. I am running on a Mac with 4 cores, and I saw that the importer was using 90%-120% CPU for when it was running.

I thought that the importer is using better compression, so I compared the sizes and the zipped layers are ~1% smaller with the importer than with gzip using default options.

Since the images, once imported, will be cached, perhaps it is not important but I thought I should let you know.

The code for trying out with command line gzip is at https://github.com/siddharthab/rules_docker/tree/container_archive

Reply:
Ack slowness, Python gzip is known to be slow, but this should be more portable than shelling out.  If the speed proves to be a problem, then we could augment `container_import` to take the uncompressed blobs through an alternate attribute and change the importer to emit those instead (should be faster since it is stored uncompressed).